### PR TITLE
Cancellable leader election

### PR DESCRIPTION
**Pull request by ash2k**
**What this PR does / why we need it**:
Adds ability to cancel leader election. Useful in integration tests where the whole app is started and stopped in each test.

**Special notes for your reviewer**:
I used the `context` package - it is impossible/hard to achieve the same behaviour with just channels without spawning additional goroutines but it is trivial with `context`. See `acquire()` and `renew()` methods.

**Release note**:

```release-note
NONE
```
/kind enhancement
/sig api-machinery

### CODE REVIEW

**Comment by ash2k** [comment]
@kubernetes/sig-api-machinery-pr-reviews PTAL

**Comment by timothysc** [comment]
@ash2k where is your integration test that makes your case?  It's pretty trivial to plumb signal-handlers through the stop channel and others have written code that depends on this external to k8s.

**Comment by ash2k** [comment]
@timothysc There seem to be some misunderstanding here. I'm saying once you start the leader election, there is no way to ask it to stop. I.e. there is no way to make the method `LeaderElector.Run()` return. It only ever returns if it fails to renew the lease. I want to start and stop integration tests that start the whole application (in a Go test) with leader election enabled (because it is code that needs to be tested). But once the application does what it needs to do in the test, the test should be able to tell the application to cleanly shutdown. No way to do it right now.
You seem to be talking about plumbing a channel though to stop the application itself - that is not the problem I'm trying to solve here.

Examples of my integration tests can be found here https://github.com/atlassian/smith/tree/master/it
You probably want to take a look at the [`SetupApp()` function](https://github.com/atlassian/smith/blob/master/it/utils_for_tests.go#L136) that starts the application for each test and then does clean tear down.

**Comment by timothysc** [comment]
I'm ok with it, but you're changing common code that will affect multiple parties that use the client.  I'm not certain how apimachinery folks communicate these changes, and what policies are in place nowadays. 

/cc @sttts

**Comment by sttts** [comment]
Would be much easier to review if the conversion from stopCh to context would be separate.

**Comment by ash2k** [comment]
It would be harder to implemented the same logic with a stop channel - channels panic if closed more than once unlike the `cancel()` function that is used to cancel a context. So I would need to have some sort of locking to ensure the channel has been closed only once.
For example the `renew()` function would need to start an additional goroutine to receive from the `stopCh` and then close a temporary channel which would signal `wait.Until()` to stop. And that temporary channel would be used to cancel the wait from inside of the `wait.Until()`. So need a lock around that - could use `sync.Once` to do this. Temporary channel is needed because it cannot and must not close the stop channel that was passed to it.
All of that is not needed with a context and it would probably be harder to review.

**Comment by sttts** [comment]
> All of that is not needed with a context and it would probably be harder to review.

I don't argue against a context :) just that we have two changes mixed into one commit here.

**Comment by sttts** [comment]
> I don't argue against a context :)

Would like us to move to contexts much more throughout the code base.

**Comment by ash2k** [comment]
I understand and do not argue that it's two changes in 1 commit. My point is that it would not be easier to review. But if you insist I'll do it in another branch as an experiment. I'll post a link.
Yes, moving from stop channels to context would be nice.

**Comment by sttts** [comment]
> But if you insist I'll do it in another branch as an experiment

Two commits in this PR would be fine.

In general, change sgtm.

**Comment by ash2k** [comment]
Done.

File: cmd/kube-scheduler/app/server.go
Comment (timothysc) [inline cmd/kube-scheduler/app/server.go]: Why is the controller manager Background and scheduler TODO contexts?

File: staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go
Comment (timothysc) [inline staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: Why are you defering a cancel the cancel should only occur then they've lost leadership.

File: staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go
Comment (timothysc) [inline staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: same

**Review by timothysc — CHANGES_REQUESTED** [review]
I'm still not sold on the overall impact.  This code is in client-go and you will be forcing this change on everyone.

File: cmd/kube-scheduler/app/server.go
Comment (ash2k) [reply cmd/kube-scheduler/app/server.go]: Because currently it blocks forever and I didn't want to change any behaviour in this pr. I guess the right thing to do would be to cancel the context when the `stop` is closed/done. To do that we'd need to start a goroutine that blocks on `stop` and cancels the context when unblocked. Will do that.

File: staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go
Comment (ash2k) [reply staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: This is a sub context, it does not influence the parent context. This code is just following the best practice patten where the resource is closed/freed when it falls out of scope.
Actually, just inspected how child cancel contexts are implemented - child contexts are referenced by parent contexts if both are cancel contexts. References are only cleaned up once the that `cancel()` function is called. So it is a good idea to clean things up.

**Comment by ash2k** [comment]
@timothysc 
> I'm still not sold on the overall impact. This code is in client-go and you will be forcing this change on everyone.

As a client-go user I can say that this is nothing compared to all the previous changes that were forced on me. And I'm not complaining, I'm actually quite happy that the code is "alive" and not forced into "let's forever maintain this api which we know is not ideal" mode. Every big breaking change was for the better. I don't see why this minor breaking change (trivial to fix - just add `context.TODO()` as the argument) is such a big deal? Exactly the same change looks like is going to be made to client apis, but on a MUCH bigger scale - https://github.com/kubernetes/community/pull/1166

**Comment by timothysc** [comment]
I'm deferring to @kubernetes/sig-api-machinery-pr-reviews from this point out.

**Review by timothysc — APPROVED** [review]

**Comment by ash2k** [comment]
Would be great to get this merged too, @sttts

**Comment by ash2k** [comment]
/assign luxas 
Ready for approval, please. @deads2k @mikedanese @luxas and someone from @kubernetes/sig-scheduling-pr-reviews

**Review by mikedanese — COMMENTED** [review]
I'm ok with these changes. If we care about API stability, we can add *WithContext functions to this package and preserve the old signatures. The package doc does say the package is alpha though.

/approve

File: cmd/cloud-controller-manager/app/controllermanager.go
Comment (ash2k) [reply cmd/cloud-controller-manager/app/controllermanager.go]: This is just future-proofing. This can be changed to use `context.TODO()` instead.

File: staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go
Comment (ash2k) [reply staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: Done

File: cmd/kube-controller-manager/app/controllermanager.go
Comment (mikedanese) [inline cmd/kube-controller-manager/app/controllermanager.go]: same nit as above. s/runCtx/ctx/

**Comment by mikedanese** [comment]
One more comment.

/lgtm
/hold

File: cmd/kube-controller-manager/app/controllermanager.go
Comment (ash2k) [reply cmd/kube-controller-manager/app/controllermanager.go]: It clashes with `ctx, err := CreateControllerContext(c, rootClientBuilder, clientBuilder, runCtx.Done())` below.

File: staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go
Comment (misterikkit) [inline staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: You could also create an overload like this:
`func (le *LeaderElector) Run() { le.Run(context.TODO()) }`

Then updating clients to use the new thing would be much more graceful.

**Review by misterikkit — COMMENTED** [review]
It would have been nicer to add new functionality to leaderelect without requiring clients to change in the same PR. Each controller could be updated in a follow-up PR so that you don't get caught up in all the approvals needed.

Since the work is already done here, I won't push for this to be split up. See my comment about how I think it could have been done, though.

File: cmd/kube-controller-manager/app/controllermanager.go
Comment (ash2k) [reply cmd/kube-controller-manager/app/controllermanager.go]: Renamed.

File: staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go
Comment (mikedanese) [reply staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: Go doesn't support overloading. We discussed adding WithContext style functions to maintain compatibility but the package is documented as having an alpha API. https://github.com/kubernetes/kubernetes/pull/57932#pullrequestreview-126079945

If people feel strongly we can do this, but I wouldn't fight for it.

File: cmd/cloud-controller-manager/app/controllermanager.go
Comment (mikedanese) [reply cmd/cloud-controller-manager/app/controllermanager.go]: If we create a subcontext, of course we should `defer cancel()` it. My question was regarding creating the subcontext in the first place.

File: staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go
Comment (misterikkit) [reply staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: Oops - I am still learning quite a bit about golang.

I am not concerned with backward compatibility, but I thought the author's life would have been easier by temporarily preserving backward compatibility.

**Comment by smarterclayton** [comment]
This is too late for 1.11 given the potential for regression.  I'm marking at as 1.12 to be explicit with that.

File: staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go
Comment (ash2k) [reply staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: @mikedanese FYI https://github.com/kubernetes/kubernetes/pull/65667

**Pull request merged**
